pkg/ebpf/c/tracee.bpf.c

```diff
@@ -963,14 +1000,76 @@ BPF_HASH(module_init_map, u32, kmod_data_t, 256);                  // holds modu
 BPF_LRU_HASH(fd_arg_path_map, fd_arg_task_t, fd_arg_path_t, 1024); // store fds paths by task
 BPF_LRU_HASH(bpf_attach_map, u32, bpf_attach_t, 1024);             // holds bpf prog info
 BPF_LRU_HASH(bpf_attach_tmp_map, u32, bpf_attach_t, 1024);         // temporarily hold bpf_attach_t
+BPF_PERCPU_ARRAY(error_scratch, bpf_error_t, 1);                   // error scratch
+BPF_HASH(errors_count, bpf_error_t, bpf_error_count_t, 4096);      // errors count
+BPF_PERCPU_ARRAY(error_count_scratch, bpf_error_count_t, 1);       // error count scratch
+BPF_PERCPU_ARRAY(error_output_scratch, bpf_error_output_t, 1);     // error output scratch
 // clang-format on
 
 // EBPF PERF BUFFERS -------------------------------------------------------------------------------
 
+BPF_PERF_OUTPUT(errors, 1024);      // errors submission
 BPF_PERF_OUTPUT(events, 1024);      // events submission
 BPF_PERF_OUTPUT(file_writes, 1024); // file writes events submission
 BPF_PERF_OUTPUT(net_events, 1024);  // network events submission
 
+// HELPERS: ERRORS ---------------------------------------------------------------------------------
+
+static __always_inline void
+do_tracee_error(void *ctx, u32 opts, enum bpf_error_id id, s64 ret, u32 line, void *file)
+{
+    if (!ctx || !file)
+        return;
+
+    u32 zero = 0;
+    bpf_error_t *err = bpf_map_lookup_elem(&error_scratch, &zero);
+    if (unlikely(err == NULL))
+        return;
+
+    err->id = id;
+    err->ret = ret;
+    err->cpu = bpf_get_smp_processor_id();
+    err->line = line;
+    READ_KERN_STR_INTO(err->file, file);
+
+    bpf_error_count_t *counter_buf = bpf_map_lookup_elem(&error_count_scratch, &zero);
+    if (unlikely(counter_buf == NULL))
+        return;
+
+    counter_buf->count = 1;
+    counter_buf->ts_prev = bpf_ktime_get_ns(); // store the current ts
+    u64 ts_prev = 0;
+
+    bpf_error_count_t *counter = bpf_map_lookup_elem(&errors_count, err);
+    if (likely(counter != NULL)) {
+        ts_prev = counter->ts_prev; // store previous ts
+
+        counter->count = counter->count + 1;
+        counter->ts_prev = counter_buf->ts_prev; // set ts_prev as current ts
+    } else {
+        counter = counter_buf;
+    }
+    bpf_map_update_elem(&errors_count, err, counter, BPF_ANY);
+
+    // submit error when its cpu occurrence time diff is greater than 2s
+    if ((counter_buf->ts_prev - ts_prev) > 2000000000) {
+        bpf_error_output_t *err_output = bpf_map_lookup_elem(&error_output_scratch, &zero);
+        if (unlikely(err_output == NULL))
+            return;
```

Comment:
since bpf_error_output_t embed bpf_error in it, we can avoid having both error_scratch and error_output_scratch maps by having only error_output_scratch and moving it up instead of error_scratch.

Reply:
There are remaining logic from the previous implementation. I'll ponder about what your statement.

Reply:
![image](https://user-images.githubusercontent.com/3372117/205156538-0550564c-c0f6-4047-9be4-972b11d66faa.png)

As discussed above, the `error_count_scratch` can be removed as we're no longer using macro functions.

But the `error_scratch` is required since it's the key of `errors_count`. For being a key, it only needs to have the unique fields to fully identify the error, which are in this case:

```c
typedef struct bpf_error {
    s64 ret;              // return value
    enum bpf_error_id id; // type
    u32 cpu;
    u32 line;                        // line number
    char file[BPF_MAX_ERR_FILE_LEN]; // filename
} bpf_error_t;
```
On the other hand, `bpf_error_output` is composed that way just to hold the output buffer with more fields (count and a required padding).

```c
typedef struct bpf_error_output {
    u32 count;
    u32 padding;
    struct bpf_error err;
} bpf_error_output_t;
```

How about the two scratches? I'm using them to diminish the stack footprint.

Reply:
Hold on. Had another coffe.

Reply:
@yanivagman you nailed it. Thanks.

![image](https://user-images.githubusercontent.com/3372117/205161112-c23de828-3e08-4f90-9b0a-dbe989d7109b.png)